torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -460,44 +452,229 @@ void CompiledFusionFunction::launch_with_tensors(at::ArrayRef<at::Tensor> inputs
+
+struct TempFile {
+  TH_DISALLOW_COPY_AND_ASSIGN(TempFile);
+  TempFile(const std::string & t, int suffix) {
+    std::vector<char> tt(t.c_str(), t.c_str() + t.size() + 1);
+    int fd = mkstemps(tt.data(), suffix);
+    JIT_ASSERT(fd != -1);
+    file_ = fdopen(fd, "r+");
+    name_ = std::string(tt.begin(), tt.end() - 1);
+  }
+  const std::string & name() const {
+    return name_;
+  }
+  void sync() {
+    fflush(file_);
+    fsync(fileno(file_));
+  }
+  void write(const std::string & str) {
+    size_t result = fwrite(str.c_str(), 1, str.size(), file_);
+    JIT_ASSERT(str.size() == result);
+  }
+  FILE* file()  {
+    return file_;
+  }
+  ~TempFile() {
+    if(file_ != nullptr) {
+      // unlink first to ensure another mkstemps doesn't
+      // race between close and unlink
+      unlink(name_.c_str());
+      fclose(file_);
+    }
+  }
+private:
+  FILE * file_ = nullptr;
+  std::string name_;
+};
+
+static void* checkDL(void * x) {
+  if(!x) {
+    barf("error in dlopen or dlsym: %s", dlerror());
+  }
+  return x;
 }
 
+struct DynamicLibrary {
+  TH_DISALLOW_COPY_AND_ASSIGN(DynamicLibrary);
+  DynamicLibrary(const char * name) {
+    handle = checkDL(dlopen(name, RTLD_LOCAL | RTLD_NOW));
+  }
+  void * sym(const char * name) {
+    JIT_ASSERT(handle);
+    return checkDL(dlsym(handle, name));
+  }
+  ~DynamicLibrary() {
+    if(!handle) return;
+    int r = dlclose(handle);
+    if(r) {
+      barf("error in dlclose: %s", dlerror());
+    }
+  }
+private:
+  void * handle = nullptr;
+};
+
+static const std::string so_template = "/tmp/pytorch_fuserXXXXXX.so";
+static const std::string cpp_template = "/tmp/pytorch_fuserXXXXXX.cpp";
+
+// TODO: have python tell us what this is...
+static const std::string compile_string =
+  "g++ -O3 -g -march=native -std=c++11 -fPIC -shared \"${cpp_file}\" -o \"${so_file}\"";
+
+static void runCompiler(const std::string & cpp, const std::string so) {
+  TemplateEnv env;
+  env.s("cpp_file",cpp);
+  env.s("so_file",so);
+  std::string result = format(compile_string,env);
+  int r = system(result.c_str());
+  JIT_ASSERT(r == 0);
+}
+
+struct CPUFusionFunction : public CompiledFusionFunction {
+  CPUFusionFunction(const std::string & name, AnnotatedGraph & agraph)
+  : CompiledFusionFunction(name, agraph)
+  , so_file(so_template, 3)
+  , cpp_file(cpp_template, 4) {
+
+    std::stringstream cu;
+    concat_desc = codegen::emitCompilationUnit(cu, name, agraph, false);
+    compilation_unit = cu.str();
+    cpp_file.write(compilation_unit);
+    cpp_file.sync();
+    runCompiler(cpp_file.name(), so_file.name());
+    so_lib.reset(new DynamicLibrary(so_file.name().c_str()));
+    kernel = reinterpret_cast<void(*)(uint32_t, void**)>(so_lib->sym(name.c_str()));
+  }
+protected:
+  virtual at::Backend backend() const override {
+    return at::kCPU;
+  }
+  virtual void launch_raw(uint32_t numel, void ** arguments) override {
+    kernel(numel, arguments);
+  }
+  TempFile so_file;
+  TempFile cpp_file;
```

Comment:
Any reason to keep those alive past the constructor by default? If something bad happens, they might not get unlinked properly.

Reply:
Hmm, Ill see if anything bad happens if we delete them early. Normally when debugging I just remove the unlinks anyway so it is not like we have to keep them around. I could imagine some debugging tools getting confused if the .so is gone.

Reply:
That's possible, but on the other hand it would be nice to not leave stuff in `/tmp` if we know we'll never be opening it anyway. I'm 99.9999% sure everything will work fine, and the real file will only be unlinked from the fs, but will still be in memory if you keep it `dlopen`ed.

Maybe store the generated source code for debugging (with some toggle)?

Reply:
It is 100% safe to unlink a file while you have a file handle open to it, the file handle stays valid. But this is only true for Linux; Windows semantics are different so that this doesn't work (the file is "busy" and not allowed to be deleted.)